### ISSUE TITLE
Please check inspection "Singleton constructor should be protected"

### ISSUE THREAD

**Issue by SergeMerge** [kalessil/phpinspectionsea#103]
Discover interesting bug in the next case:
![image](https://cloud.githubusercontent.com/assets/14336083/21807456/757ad646-d747-11e6-9621-0c25bc16ace0.png)
Also, this warning should appear when _protected $instance_ is static. This thing very annoying.

**Comment by kalessil** [comment]
Thank for reporting @SergeMerge , is the complete components code available anywhere (I'm interested in methods it provides)?

Annoying in general or probably some documentation with options how to solve would make it less annoying?

**Comment by SergeMerge** [comment]
Hello @kalessil , i cant provide full code of this. This warning only confusing me, because $instance field not a static. I never saw cases (in PHP), where singleton pattern can be implemented with non-static $instance field.

**Comment by SergeMerge** [comment]
Code example:

```
class ExampleBuilder implements BuilderInterface
{
    private $instance;

    /** @var EntityRepository */
    private $repository;

    /**
     * CurrencyPairRatingBuilder constructor.
     * @param EntityRepository $repository
     */
    public function __construct(EntityRepository $repository)
    {
        $this->repository = $repository;
    }

    /**
     * @param array $data
     *
     * @return mixed
     */
    public function build(array $data)
    {
        $this->instance = new Entity();
        return $this;
    }

    /**
     * @return mixed
     */
    public function getInstance()
    {
        return $this->instance;
    }
}
```

**Comment by rentalhost** [comment]
I guess that a solution is instead of check if `$instance` was declared as a property to mark it as a possible Singleton, check if any property is used to store a static new instance of current class and the same method returns it. Then things like that should be reported:

```php
class Test {
    private $property;

    public function getProperty() {
        if(!$this->property) {
            $this->property = new static;
        }

        return $this->property;
    }
}
```

**Comment by kalessil** [comment]
@rentalhost : the example is a missleading interface definition.

@SergeMerge : the example actually belongs to let say misleading interface, Builder design pattern looks a little bit different. But getting closer to real-life, I suggesting 2 option which you probably can apply to the original class:
- returning Entity from the build (=> that's the Builder pattern definition)
- renaming getInstance to getBuiltEntity/getResult (=> fixing lack of semantic)

I also convinced that the current version of inspection did it's job and spotted a good example, hence the behavior remains unchanged.
